tests/diagnostics/ctor-retval-legal.slang

```diff
@@ -0,0 +1,82 @@
+//TEST(compute):COMPARE_COMPUTE(filecheck-buffer=CHECK): -shaderobj -output-using-type
+//TEST(compute):COMPARE_COMPUTE(filecheck-buffer=CHECK): -vk -shaderobj -output-using-type
+
+//TEST_INPUT:ubuffer(data=[0 0 0 0 0 0 0 0 0 0 0 0], stride=4):out,name=outputBuffer
+
+// Test to catch errors with return values in constructors within an ordinary struct.
+// A constructor from the callsite's point of view is a function that returns the struct type.
+// A constructor from the inside the body is treated as a function that modifies `this` and not return.
+
+// In the following test-cases where the constructor either returns an expr or no value,
+// within a condition for early exit or not, it automatically gets resolved as
+// this = t;
+// return *this; or return;
+RWStructuredBuffer<float> outputBuffer;
+
+interface IFoo {
+    __init();
+    __init(int);
+    __init(float);
+    void compute(int);
+}
+
+struct Impl : IFoo
+{
+    float x;
+
+    __init(float v) {
+        if (v > 0) {
+            this.x = 2 * v;
+            return this;
+        }
+        else {
+            this.x = 3 * v;
+            return;
+        }
+    }
+
+    __init(int ival) {
+        float val = ival;
+        Impl v = Impl(val);
```

Comment:
What is Impl?
Is it same as "This"?

Reply:
Yes. `This` is a typedef for the owning type. So `Impl` and `This` should mean the same thing in this context.